src/device/arm/arm.go

```diff
@@ -169,3 +219,25 @@ func SystemReset() {
 		Asm("wfi")
 	}
 }
+
+
+func SetupSystemTimer(cycle_count uint32) error {
```

Comment:
The function needs some GoDoc comments, please.

Also, the parameter should not use snake case. 

I wonder if it would be good for the param to this function was a `time.Duration` that would then be converted to the correct number of cycles in the function's implementation?

Reply:
> I wonder if it would be good for the param to this function was a `time.Duration` that would then be converted to the correct number of cycles in the function's implementation?

Yeah, the setup function in https://gist.github.com/Infinoid/ea2002fbf6989f8dbaff57a63ccda2ba did that.  I removed it for a few reasons:

1.  circular import problems.  When you import "time" in `device/arm/arm.go`, you get:

```
error: import cycle:
	runtime
	device/arm
	time
	runtime
```

2.  This interface assumes that the CPU frequency never changes.  The system timer is based on the CPU clock rate, so if the CPU enters a lower power state, your interrupts won't fire at the duration you specify.  When the setup API asks for a cycle count, this is obvious to the user.  When the setup API asks for a time.Duration, this is not obvious.

3.  Sometimes the hardware provides a 10ms reference count in SYST_CALIB_TENMS, sometimes it doesn't.  (`atsamd51` doesn't.)  If it doesn't, you can use machine.CPU_FREQUENCY to do the calculation.  That constant is not available to `src/device/arm/arm.go`, due to another circular import problem.

4.  Precision issues.  I used floating point division, and the results were not precise.  I tried telling to run every 100ms.  The tick period I actually got was 99.8ms.  Close, but cycle counts can do better.

So, maybe we can make it work?  But passing a cycle count is much simpler.

Reply:
> The function needs some GoDoc comments, please.
> 
> Also, the parameter should not use snake case.

Done.

Reply:
Looks good.